File: tests/cpu/julia/test_manipulation.jl
# Manipulation CPU tests — aligned with C++ test_manipulation.cpp
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu julia tests/cpu/julia/test_manipulation.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

passed = 0; failed = 0
function check(name, cond)
    global passed, failed
    if cond; passed += 1; else; failed += 1; println("FAIL: $name"); end
end

println("=== Manipulation ===")

# reshape
a = Insight.arange(0.0, 6.0, 1.0, Insight.float64)
b = Insight.reshape(a, Int64[2, 3])
check("reshape numel", Insight.numel(b) == 6)
check("reshape ndim", Insight.ndim(b) == 2)

# transpose
a = Insight.from_data(Float64[1 2 3; 4 5 6])
b = Insight.transpose(a)
check("transpose numel", Insight.numel(b) == 6)

# permute
a = Insight.ones(Int64[2, 3, 4], Insight.float64)
b = Insight.permute(a, Int32[2, 0, 1])
check("permute", Insight.numel(b) == 24)

# swapaxes
a = Insight.ones(Int64[2, 3], Insight.float64)
b = Insight.swapaxes(a, 0, 1)
check("swapaxes", Insight.numel(b) == 6)

# moveaxis
a = Insight.ones(Int64[2, 3, 4], Insight.float64)
b = Insight.moveaxis(a, 0, 2)
check("moveaxis", Insight.numel(b) == 24)

# fliplr
m2 = Insight.from_data(Float32[1 2 3; 4 5 6])
check("fliplr", Insight.fliplr(m2).ptr != C_NULL)

# flipud
check("flipud", Insight.flipud(m2).ptr != C_NULL)

# rot90
a2 = Insight.from_data(Float64[1 2; 3 4])
b = Insight.rot90(a2)
check("rot90", Insight.numel(b) == 4)

# diag_fn
d = Insight.from_data([1.0, 2.0, 3.0])
b = Insight.diag_fn(d)
check("diag_fn numel", Insight.numel(b) == 9)
check("diag_fn ndim", Insight.ndim(b) == 2)

# diagonal
a3 = Insight.from_data(Float64[1 2 3; 4 5 6; 7 8 9])
b = Insight.diagonal(a3)
check("diagonal", Insight.numel(b) == 3)

# tril
check("tril", Insight.tril(m2).ptr != C_NULL)

# triu
check("triu", Insight.triu(m2).ptr != C_NULL)

# diff_fn
a4 = Insight.from_data([1.0, 3.0, 6.0, 10.0])
b = Insight.diff_fn(a4)
check("diff_fn", Insight.numel(b) == 3)

# squeeze (via reshape - Julia doesn't have direct squeeze)
a = Insight.zeros(Int64[1, 3, 1], Insight.float64)
# squeeze is not directly available in Julia bindings
# Use reshape as alternative
b = Insight.reshape(a, Int64[3])
check("squeeze_via_reshape", Insight.numel(b) == 3)

println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0; exit(1); end

File: tests/cpu/python/test_manipulation.py
"""Manipulation CPU binding tests — aligned with C++ test_manipulation.cpp.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python
    LD_LIBRARY_PATH=build/backends/cpu
    python -m pytest tests/cpu/python/test_manipulation.py -v
"""

import sys, os, pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)


class TestManipulationCPU:
    """Array manipulation operations — CPU backend."""

    def test_reshape(self):
        a = ins.from_numpy(np.arange(6, dtype=np.float64))
        b = ins.reshape(a, [2, 3])
        assert b.numel() == 6
        assert b.ndim() == 2

    def test_transpose(self):
        a_np = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float64)
        b = ins.permute(ins.from_numpy(a_np), [1, 0])
        np.testing.assert_allclose(b.numpy(), a_np.T)

    def test_squeeze(self):
        a = ins.from_numpy(np.zeros([1, 3, 1], dtype=np.float64))
        b = ins.squeeze(a)
        assert b.numel() == 3

    def test_unsqueeze(self):
        a = ins.from_numpy(np.zeros([3], dtype=np.float64))
        b = ins.unsqueeze(a, 0)
        assert b.ndim() == 2

    def test_concat(self):
        a = ins.from_numpy(np.array([1, 2], dtype=np.float64))
        b = ins.from_numpy(np.array([3, 4], dtype=np.float64))
        c = ins.concat([a, b])
        assert c.numel() == 4
        np.testing.assert_allclose(c.numpy(), [1, 2, 3, 4])

    def test_concat_axis(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        b_np = np.array([[5, 6]], dtype=np.float64)
        c = ins.concat([ins.from_numpy(a_np), ins.from_numpy(b_np)], axis=0)
        np.testing.assert_allclose(c.numpy(), np.concatenate([a_np, b_np], axis=0))

    def test_split(self):
        a = ins.from_numpy(np.arange(6, dtype=np.float64))
        parts = ins.split(a, 3, axis=0)
        assert len(parts) == 3
        assert parts[0].numel() == 2

    def test_tile(self):
        a = ins.from_numpy(np.array([1, 2, 3], dtype=np.float64))
        b = ins.tile(a, [2])
        assert b.numel() == 6

    def test_repeat(self):
        a = ins.from_numpy(np.array([1, 2, 3], dtype=np.float64))
        b = ins.repeat(a, 2)
        assert b.numel() == 6

    def test_flip(self):
        a_np = np.array([1, 2, 3, 4], dtype=np.float64)
        result = ins.flip(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.flip(a_np))

    def test_roll(self):
        a_np = np.array([1, 2, 3, 4, 5], dtype=np.float64)
        result = ins.roll(ins.from_numpy(a_np), 2)
        np.testing.assert_allclose(result.numpy(), np.roll(a_np, 2))

    def test_pad(self):
        a = ins.from_numpy(np.array([1, 2, 3], dtype=np.float64))
        b = ins.pad(a, [2, 3])
        assert b.numel() == 8

    def test_diag(self):
        a = ins.from_numpy(np.array([1, 2, 3], dtype=np.float64))
        b = ins.diag(a)
        assert b.numel() == 9
        assert b.ndim() == 2

    def test_tril(self):
        a = ins.from_numpy(np.ones([3, 3], dtype=np.float64))
        b = ins.tril(a)
        np.testing.assert_allclose(b.numpy(), np.tril(np.ones([3, 3])))

    def test_triu(self):
        a = ins.from_numpy(np.ones([3, 3], dtype=np.float64))
        b = ins.triu(a)
        np.testing.assert_allclose(b.numpy(), np.triu(np.ones([3, 3])))

    def test_diff(self):
        a_np = np.array([1, 3, 6, 10], dtype=np.float64)
        result = ins.diff(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.diff(a_np))

    def test_swapaxes(self):
        a = ins.from_numpy(np.ones([2, 3, 4], dtype=np.float64))
        b = ins.swapaxes(a, 0, 2)
        assert b.numel() == 24

    def test_moveaxis(self):
        a = ins.from_numpy(np.ones([2, 3, 4], dtype=np.float64))
        b = ins.moveaxis(a, 0, 2)
        assert b.numel() == 24

    def test_fliplr(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        result = ins.fliplr(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.fliplr(a_np))

    def test_flipud(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        result = ins.flipud(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.flipud(a_np))

    def test_rot90(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        result = ins.rot90(ins.from_numpy(a_np))
        np.testing.assert_allclose(result.numpy(), np.rot90(a_np))

    def test_flatten(self):
        a = ins.from_numpy(np.ones([2, 3], dtype=np.float64))
        b = ins.flatten(a)
        assert b.numel() == 6
        assert b.ndim() == 1

    def test_ravel(self):
        a = ins.from_numpy(np.ones([2, 3], dtype=np.float64))
        b = ins.ravel(a)
        assert b.numel() == 6
        assert b.ndim() == 1


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
